File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:        
        def generateHelper(left_cnt, right_cnt, cur_str):
            if len(cur_str) == n * 2:
                res.append(cur_str)
                
            if left_cnt < n:
                generateHelper(left_cnt + 1, right_cnt, cur_str + "(")
                
            if right_cnt < left_cnt:
                generateHelper(left_cnt, right_cnt + 1, cur_str + ")")
               
        res = []
        generateHelper(0, 0, "")
        
        return res
                
        

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        int left_count = 0, right_count = 0;
        vector<string> res;
        helper(res, "", n, left_count, right_count);
        return res;
    }

private:
    void helper(vector<string> &res, string cur, int n, int left, int right) {
        if (left == n && right == n) {
            res.push_back(cur);
            return;
        }

        if (left < n) {
            helper(res, cur + "(", n, left + 1, right);
        }

        if (right < left) {
            helper(res, cur + ")", n, left, right + 1);
        }
    }
};